malai-core/org.malai.ts/src/binding/AnonNodeBinding.ts

```diff
@@ -122,12 +127,11 @@ export class AnonNodeBinding<C extends CommandImpl, I extends TSInteraction<D, F
 
     public fsmStops(): void {
         super.fsmStops();
-        if (this.currentCmd !== undefined) {
-            this.endOrCancelFct(this.getInteraction().getData(), this.currentCmd);
+        if (this.endOrCancelFct !== undefined) {
+            this.endOrCancelFct(this.getInteraction().getData(), this.cmd);
         }
-        if (this.currentCmd !== undefined) {
-            this.onEnd(this.getInteraction().getData(), this.currentCmd);
+        if (this.onEnd !== undefined) {
+            this.onEnd(this.getInteraction().getData(), this.cmd);
         }
-        this.currentCmd = undefined;
```

Comment:
why suppressing the this.cmd = undefined
?

Reply:
Because in super.fsmStops(), this.cmd is always set to undefined

Reply:
so, now in this method `this.cmd` is always `undefined` so `this.onEnd(this.getInteraction().getData(), this.currentCmd);` will also give `undefined` as a second argument.

I remember now why I added `currentCmd` in this class: to overcome this problem. The attribute `currentCmd` is a mirror of `cmd` to avoid having `undefined` after a super call to `onEnd` or `onCancel`.
So, you cannot suppress `currentCmd` :s

Reply:
I check currentCmd and it was never set, always defined to undefined.
This is why end() routine never run

Reply:
You found the problem ;)
In JavaFX `currentCmd` is set in:
```java
	@Override
	protected C map() {
		if(cmdProducer == null) {
			currentCmd = super.map();
		}else {
			currentCmd = cmdProducer.apply(getInteraction().getData());
		}
		return currentCmd;
	}
```
This method is not present in the TS version.
I think the solution is to add this method in this TS class.

Reply:
I can add a saveCmdForStop in WidgetBindingImpl to get around the issue.
In fsmStop this.cmd will be store in, and I add a getter to be use in AnonNodeBinding

like this : 
```typescript

// in WidgetBindingImpl

private saveCmdForStop: C | undefined

// in fsmStop

{
this.saveCmdForStop = this.cmd;
this.cmd = undefined;
}

public getSaveCmd() { return this.saveCmdForStop }

// in AnonNodeBinding

if (this.onEnd !== undefined and this.getSavedCmd() !== undefined)

```

Reply:
I prefer to have the implementation in JavaFX and TS